### ISSUE TITLE
PPA install of 0.14 on RPI

### ISSUE THREAD

**Issue by 3kay** [learningequality/ka-lite#4405]
Hello,

I have installed 0.14 via PPA on a raspberry pi 2 as per installation guide with only one tweaking.
While the repository was properly loaded, apt-get update gave me 404 error of not found on ka-lite.
I had to manually check the /etc/apt/sources.list.d/learningequality-ka-lite-wheezy.list and changed both the deb and deb-src files to refer to /ubuntu trusty main instead of /ubuntu wheezy main which were not found. I don't know if it was right to do so but the installation went perfectly alright on sudo apt-get install ka-lite-raspberry-pi 
Now I am having a strange problem.
At boot, I get the usual message to access ka-lite
from other systems with
http://192.168.1.10:7007/
http://1.1.1.1:7007/ 
from this machine
http://127.0.0.1:7007/

7007 instead of 8008

The last message says
My IP address is 192.168.1.10 1.1.1.1 169.254.94.57

The first two addresses are understandable but where did the THIRD ADDRESS come from?
It is not even a range in our country!
I can get the interface of ka-lite but can't register the device.
It says to refresh when connected to the internet but it is surly connected to the internet. I can ping google by name in command line.
The exercises are available but when i try to download any video in any subject it does say successfully launched but then quickly exits with a flash message 'refused connection' that vanishes very quickly.
Is this all because i used the wrong name in PPA? 
I have a static IP DSL line but its DHCP allocates local IPs to the attached systems. Any thing to do with NATing? I am using a static local IP in the /etc/network/interfaces  
The Epiphany browser does not go to the Internet but the whole system was installed while connected with a lan cable.
Any hint would be highly appreciated.

Thanking in anticipation,

Best,

Lutfullah

**Comment by 3kay** [comment]
PPA installation of 0.14 on Ubuntu in a laptop is working fine.  So, i assume it was the non wheezy ppa that was causing the problems in a raspberry pi installation. I do not know where to find the PPA repository for the wheezy version of 0.14 or any other version. Google does not seem to be helping in this search. Could someone please point me to the proper PPA repository? I know this may be the wrong place to ask such questions but there is very little activity on the forum that I am subscribed to. LKK

**Comment by MCGallaspy** [comment]
Hi @3kay, looks like you've got several different issues/questions here. First of all, 127.0.0.1 is a [reserved IP address](https://en.wikipedia.org/wiki/Reserved_IP_addresses) that just refers to your own machine.

Perhaps @benjaoming and @aronasorman can comment on the following issues:
* Port 7007 displayed instead of 8008
* Installing using the PPA on Ubuntu. Sounds like you're using version 14.04 (Trusty Tahr), is that correct?

@jamalex, can you elaborate on why @3kay is unable to register his/her device, even though he/she is connected to the internet? Could that be related to the inability to download videos? Might it be due to the outage of the central server earlier? It looks like raspberry pi settings use the same central server.

@3kay can you see if hub.learningequality.org is reachable from your machine? You could receive the error you described if it is not.

Thanks for reporting your issues, and we'll certainly try to address them when we're able.

**Comment by benjaoming** [comment]
@3kay KA Lite for raspberry pi installs nginx and you have to access it on port 8008. Nginx forwards page serving to ka-lite, but static media (images, videos etc) are served by nginx because it's less memory intensive.

Regarding the third address, try looking at output from `ifconfig`, perhaps that will tell you about a third network interface.

**Comment by jamalex** [comment]
> @jamalex, can you elaborate on why @3kay is unable to register his/her device, even though he/she is connected to the internet?

@3kay, can you describe your network setup (in particular, how the pi is connected to the net, and how your client browser is connected to the pi -- or is it on the pi itself)? It's possible that even though the Pi has internet, your client device connecting to the Pi can't connect to the internet, which could cause issues.

> Regarding the third address, try looking at output from ifconfig, perhaps that will tell you about a third network interface.

169.x addresses are typically internal/private addresses that can indicate no DHCP server could be found (e.g. see http://www.tech-101.com/support/topic/1100-whats-that-169xxx-address/).

**Comment by 3kay** [comment]
@MCGallaspy,  @benjaoming , @jamalex  I am so glad to see your responses. After failing to have a successful installation using PPA ubuntu trusty and raspbian OS on rpi, I tried to install Ubuntu 14.04 LTS ( https://wiki.ubuntu.com/ARM/RaspberryPi) on my raspberry pi and then ppa installed ka-lite. The installation went smooth like before but ended up in exactly the same situation. :7007 instead of :8008 and no registration or video download. 

At least one problem got remedied when I re-installed Raspbian wheezy on rpi and used the main method of debian bundle pkg install.

I am getting now the :8008 setup and :7007 has disappeared (although I do not know what is the difference).

The ifconfig does show an extra mon.wlan0 which may explain the private IP 169.**** assignment.

 eth0      Link encap:Ethernet  HWaddr b8:27:eb:8e:43:e2  
          inet addr:192.168.1.7  Bcast:192.168.1.255  Mask:255.255.255.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:162 errors:0 dropped:4 overruns:0 frame:0
          TX packets:109 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:9446 (9.2 KiB)  TX bytes:8798 (8.5 KiB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:1361 errors:0 dropped:0 overruns:0 frame:0
          TX packets:1361 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:2221036 (2.1 MiB)  TX bytes:2221036 (2.1 MiB)

mon.wlan0 Link encap:UNSPEC  HWaddr C8-3A-35-CA-80-35-00-00-00-00-00-00-00-00-00-00  
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:767 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:77104 (75.2 KiB)  TX bytes:0 (0.0 B)

wlan0     Link encap:Ethernet  HWaddr c8:3a:35:ca:80:35  
          inet addr:1.1.1.1  Bcast:1.1.1.255  Mask:255.255.255.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:65 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:14835 (14.4 KiB)

The /etc/network/interfaces file is probably the one which is giving problems.

auto lo
iface lo inet loopback

auto eth0
#allow-hotplug eth0
iface eth0 inet manual

auto wlan0
#allow-hotplug wlan0
#wpa-conf /etc/wpa_supplicant/wpa_supplicant.conf      <<<<<<

auto wlan1
#allow-hotplug wlan1
iface wlan1 inet manual
wpa-conf /etc/wpa_supplicant/wpa_supplicant.conf

iface wlan0 inet static
  address 1.1.1.1
  netmask 255.255.255.0
  pre-up sudo iptables -t nat -A PREROUTING -i wlan0 -p tcp -j DNAT --to-destination 1.1.1.1
  pre-up sudo python /opt/ka-lite-pi-scripts/redirect_port80_to_port8008.py &

One of the line "#wpa....." marked with "<<<<<<< " was hashed out by me. It was causing read failure error of /etc/network/interfaces at the time of boot, leading to complete failure. 
I do not know if it is required. 
The wlan1 is also redundant looking.
The lower 5 lines were probably written by:
python ./configure_network_interfaces.py script.

I can ping your hub:
PING hub.learningequality.org (1.1.1.1) 56(84) bytes of data.
64 bytes from 3kay3.local (1.1.1.1): icmp_req=1 ttl=64 time=0.165 ms
64 bytes from 3kay3.local (1.1.1.1): icmp_req=2 ttl=64 time=0.197 ms
64 bytes from 3kay3.local (1.1.1.1): icmp_req=3 ttl=64 time=0.202 ms
64 bytes from 3kay3.local (1.1.1.1): icmp_req=4 ttl=64 time=0.201 ms
64 bytes from 3kay3.local (1.1.1.1): icmp_req=5 ttl=64 time=0.180 ms
64 bytes from 3kay3.local (1.1.1.1): icmp_req=6 ttl=64 time=0.274 ms

--- hub.learningequality.org ping statistics ---
6 packets transmitted, 6 received, 0% packet loss, time 5008ms
rtt min/avg/max/mdev = 0.165/0.203/0.274/0.035 ms

Shouldn't the ip in parenthesis (1.1.1.1)  be something other?  Don't know really.

There is nothing wrong with my Internet connection. I used it for ka-lite installation of 0.2 on a raspberry pi and that is still up and running, i am told.

The PPA problems seems to be related to nginx configuration. During my experiments i saw a couple of times that the browser would show the default page of nginx when an address like google.com was typed in the browser. Also, in the nginx configuration files i saw two soft links which were pointing back to their selves. I have no idea but they looked odd to me. (sorry i did not take a note of the link names). 

I am not going to experiment further with re-installations but would keep trying to resolve this issue with current setup.

Thank you all once again for your time and help and more for the noble work you are doing for humanity,

Lutfullah

**Comment by benjaoming** [comment]
@3kay For the raspberry pi version of ka-lite, nginx listens on 8008, and the ka-lite gateway is listening on 7007, but you should access the server on 8008.

If you installed the `ka-lite` package from the PPA, you'll have a package that does not configure nginx and thus uses more resources for serving static media. You should install `ka-lite-raspberry-pi` instead. Before doing that, remove your `.kalite/settings.py` and do `sudo apt-get purge ka-lite` to completely remove the other package.

**Comment by 3kay** [comment]
@benjaoming right now I have only one problem. I am using ka-lite on raspberry pi and the raspberry pi browser on the same machine. I can ping anywhere and the Internet is working. Yet when I try to register the device - it says the device needs to have an Internet connection and is not registering the device. I can't download a video also.

**Comment by 3kay** [comment]
@benjaoming I re-installed ka-lite from the -bundle.deb file not ppa.

**Comment by 3kay** [comment]
@MCGallaspy , @benjaoming , @jamalex 

Up and running with PPA fresh install on another SD card. Managed port settings with ka-lite-gtk. I can now register the device as well as download videos.   
Thank you so much,

Lutfullah

**Comment by benjaoming** [comment]
@3kay thanks for updating us on your solutions!

 Regarding this:

> Yet when I try to register the device - it says the device needs to have an Internet connection and is not registering the device. I can't download a video also. 

I'm pretty sure we had an issue that was solved related to this behaviour, maybe @MCGallaspy can find it, because I can't :)

**Comment by MCGallaspy** [comment]
Hi @3kay, glad you resolved your issue, but I wonder if you can share any more details on how you resolved it for the benefit of other people that might wander in.

>I'm pretty sure we had an issue that was solved related to this behaviour, maybe @MCGallaspy can find it, because I can't :)

In that issue, users were redirected away from the video download page. @3kay when you navigated to the video download page, were you redirected away from it? Or were you able to navigate to the page, but unable to download videos? If the latter, was there an error message?

> Up and running with PPA fresh install on another SD card. Managed port settings with ka-lite-gtk. I can now register the device as well as download videos.

What did you do differently this time to resolve the issue? Can you list the steps, so that someone else could try the same thing?

**Comment by 3kay** [comment]
@MCGallaspy 
About the video download, this is what I wrote in my first post:
"-----when i try to download any video in any subject it does say successfully launched but then quickly exits with a flash message 'refused connection' that vanishes very quickly"
Surely i was navigated to the download page - the topic tree also came on screen - after selection of a particular video - it gave a message -- refused  connection that disappeared quickly. I wouldn't count it as having anything to do on your side because my installation was having problems. Particularly the addresses and ports which were bound to port 7007. 

I dd transferred the problematic installation to another sd card for exploration. Some how i had doubts about nginx giving problems. I completely removed/purged nginx - including even rm -rf of its  /etc directory and re-installed nginx. After that,  I re-installed ka-lite-raspberry-pi from ppa repository. I ended up with exactly the same situation with an installation in which all addresses were bound to :7007. I therefore installed ka-lite-gtk for changing the port. On a raspberry pi - this gui can be installed only if you invoke startx and install it from the command line window available inside the x.
Even the ka-lite-gtk gui refused to change the port from 7007 to 8008. It required authentication with a real root account and not just sudoer. The port got changed to 8008 and so did everything else. 
I have still to set the wlan for wireless access.
I might write a blog about the whole setup which will carry all the steps required for having a running ka-lite server on a raspberry pi. 
Sorry about closing the issue - i thought it was over.

**Comment by 3kay** [comment]
@MCGallaspy, @benjaoming, @jamalex   I have now two almost similar PPA installs of 0.14 for one raspberry pi.

The first setup 3kay3 had no problem before. I registered the device and downloaded a test video also. Today, I configured the access point for wireless access. It went well with no problem but now I can't login with the superuser account on this setup using the browser of raspberry pi itself.

The second, 3kay4, gave me the same problem of 7007 vs 8008 ports but I finally got these changed to 8008 with ka-lite-gtk. This one, although connected to the Internet,  is refusing to register the device - and no video downloads are possible. This setup too is refusing to login with the superuser account. I created a new superuser account but that was also refused.
During installation of 3kay4, I observed something that I would like to share:
The installation was binding all addresses to 7007 while reporting that 8008 is used by some other service. I tried to check and found that it was used by no other service but the nginx itself.
it was showing tcp ....0.0.0.0:8008 ....xxxx/xxx nginx
Also, in the boot messages, i found: 
nginx:[emerg] bind () to 0.0.0.0:80 failed (98: Address already in use )
By the end of the week, hopefully, I will have a write up of all the steps I am doing.
I am sure these problems will be over and the write up can then be updated accordingly.

Thank you all,

lutfullah
